---
title: "SIMD Tutorial"
author: "Yan Zhou"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{SIMD}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

#1 Introduction
SIMD method is developed for jointly analyzing MeDIP-seq and MRE-seq data, 
which are derived from methylated DNA immunoprecipitation (MeDIP) experiments
followed by sequencing (MeDIP-seq) and methyl-sensitive restriction enzymes
experiments for unmethylated CpGs (MRE-seq). We have implemented the
SIMD method via a set of R functions with the computational intensive parts
written in C. We make a R package named SIMD and give a tutorial for
the package. The method consist three steps.

Step 1: Data Pre-preparation;
    Calculate the CpG count.
    Calculate the MRE CpG count.
    Calculate MeDIP-seq tag count of each site of control and treatment samples.
    Calculate MRE-seq tag count of each site of control and treatment samples.

Step 2: Calculating p-values of each CpG site by the EM test;

Step 3: q-values for FDR control.

We use a part of a real dataset to illustrate the usage of the SIMD package.
The programs can run under the windows 10 system. The
R versions should larger than 3.5.0.

#2 Preparations
Before installing SIMD package, the user must install two other R pack-
ages, which can be done using the following commands.
```{r, eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("edgeR")
biocLite("statmod")
biocLite("methylMnM")
```

```{r}
library(edgeR)
library(statmod)
library(methylMnM)
```
Next, to install the SIMD package into your R environment, start R and
enter:
```{r, eval=FALSE}
biocLite("SIMD")
```
Then, the SIMD package is ready to load.
```{r}
library(SIMD)
```

#3 Data format
In order to reproduce the presented SIMD workflow, the package includes the 
example data sets, which is a part of a real data sets, including: 
all_CpGsite_bin_chr18.RData,

three_mre_cpg.RData, 

EM2_H1ESB1_MeDIP_sigleCpG.RData, 

EM2_H1ESB2_MeDIP_sigleCpG.RData,

H1ESB1_MRE_sigleCpG.RData, 
and H1ESB2_MRE_sigleCpG.RData in the data subdirectory of the SIMD package.
The files contain genomic regions from chromosome 18 only, as covered by
short reads obtained from a MeDIP and MRE experiment of human H1ESB1
cells.

#4 Data Pre-processing
The SIMD program requires users to provide the genome-wide MeDIP-seq
and MRE-seq reads and the reference genome. The pre-processing step involves
binning and the calculation of MeDIP-seq and MRE-seq count and CpG and
MRE-specific count in each CpG site. The mainly steps please refer the data
pre-processing steps of R package MethylMnM. 

In this manual, we use example data located in the data subdirectory of the 
SIMD package, and the example data is pre-processed by MethylMnM package in 
advance. 

```{r,eval=FALSE}
datafile<-system.file("extdata", package = "methylMnM")
data(example_data)

filepath<-datafile[1]
allcpgfile<-EM_H1ESB1_MeDIP_sigleCpG

dirwrite<-paste(setwd(getwd()),"/",sep="")
readshort<-paste(filepath,"/H1ESB1_Medip_18.extended.txt",sep="")
writefile<-paste(dirwrite,"EM2_H1ESB1_MeDIP_sigleCpG.bed",sep="")
reportfile<-paste(dirwrite,"EM2_H1ESB1_MeDIP_sigleCpG_report.bed",sep="")
f<-EMalgorithm(cpgsitefile=readshort,allcpgfile=allcpgfile,category="1",
writefile=writefile,reportfile=reportfile)
```

The CpG count, MRE-CpG count, MeDIP-seq count and MRE-seq count of
each bin are stored at
```{r}
dirwrite<-paste(setwd(getwd()),"/",sep="")
```

#5 p-values of SIMD test
In order to detect different methylated CpG sites, we should calculate p-value 
of each site. The below codes are calculate p-value of each site with SIMD 
method. The input files which include "datafile", "cpgfile" and "mrecpgfile" 
are should have been generated by Step 1. The output file "writefile" will own 
eleven columns, that is, "chr", "chrSt","chrEnd","Medip1","Medip2","MRE1",
"MRE2", "cg","mrecg","pvalue",'Ts'. We also output a reportfile which will 
include parameters "s1/s2"; "s3/s4"; "N1"; "N2"; "N3"; "N4"; "c1"; "c2"; 
and "Spend time".


```{r}
data(example_data)
data1<-EM2_H1ESB1_MeDIP_sigleCpG
data2<-EM2_H1ESB2_MeDIP_sigleCpG
data3<-H1ESB1_MRE_sigleCpG
data4<-H1ESB2_MRE_sigleCpG
datafile<-cbind(data1,data2,data3,data4)
allcpg<-all_CpGsite_bin_chr18
mrecpg<-three_mre_cpg
dirwrite<-paste(setwd(getwd()),"/",sep="")

writefile<-paste(dirwrite,"pval_EM_H1ESB1_H1ESB21.bed",sep="")
reportfile<-paste(dirwrite,"report_pvalH1ESB1_H1ESB21.bed",sep="")
```

```{r}
EMtest(datafile=datafile,chrstring=NULL,cpgfile=allcpg,
mrecpgfile=mrecpg,writefile=writefile,reportfile=reportfile,
mreratio=3/7, psd=2,mkadded=1,f=1)
```
